File: frameworks/php-laravel/tests/Feature/MutationTests.php
<?php

namespace Tests\Feature;

use Tests\TestCase;
use Tests\Traits\DatabaseTrait;
use App\Models\User;
use App\Models\Post;

/**
 * MutationTests - Tests PUT/PATCH endpoints and state changes
 *
 * Note: These tests verify that mutations (update operations) work correctly
 * by directly updating the models and verifying the changes persist.
 *
 * Coverage includes:
 * - Single field updates
 * - Multi-field updates
 * - Immutable field protection
 * - State change verification
 * - Input validation
 */
class MutationTests extends TestCase
{
    use DatabaseTrait;

    // ========================================================================
    // Single Field Update Tests
    // ========================================================================

    public function test_update_user_bio_single_field(): void
    {
        // Arrange
        $user = $this->createTestUser('alice', 'alice@example.com', 'Alice', 'Old bio');
        $userId = $user->id;
        $newBio = 'Updated bio';

        // Act
        $user->update(['bio' => $newBio]);

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson(['bio' => $newBio]);

        // Verify other fields unchanged
        $response->assertJson(['fullName' => 'Alice']);
    }

    public function test_update_user_full_name_single_field(): void
    {
        // Arrange
        $user = $this->createTestUser('bob', 'bob@example.com', 'Bob');
        $userId = $user->id;
        $newName = 'Bob Smith Updated';

        // Act
        $user->update(['full_name' => $newName]);

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson(['fullName' => $newName]);
    }

    public function test_update_post_title(): void
    {
        // Arrange
        $author = $this->createTestUser('author', 'author@example.com', 'Author');
        $post = $this->createTestPost($author->pk_user, 'Original Title', 'Original Content');
        $postId = $post->id;
        $newTitle = 'Updated Title';

        // Act
        $post->update(['title' => $newTitle]);

        // Verify
        $response = $this->getJson("/api/posts/{$postId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson(['title' => $newTitle])
            ->assertJson(['content' => 'Original Content']);
    }

    public function test_update_post_content(): void
    {
        // Arrange
        $author = $this->createTestUser('author', 'author@example.com', 'Author');
        $post = $this->createTestPost($author->pk_user, 'Title', 'Original Content');
        $postId = $post->id;
        $newContent = 'Updated Content';

        // Act
        $post->update(['content' => $newContent]);

        // Verify
        $response = $this->getJson("/api/posts/{$postId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson(['content' => $newContent]);
    }

    // ========================================================================
    // Multi-Field Update Tests
    // ========================================================================

    public function test_update_user_multiple_fields(): void
    {
        // Arrange
        $user = $this->createTestUser('charlie', 'charlie@example.com', 'Charlie');
        $userId = $user->id;
        $newBio = 'New bio';
        $newName = 'Charlie Updated';

        // Act
        $user->update([
            'bio' => $newBio,
            'full_name' => $newName,
        ]);

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson([
                'bio' => $newBio,
                'fullName' => $newName,
            ]);
    }

    public function test_update_post_title_and_content(): void
    {
        // Arrange
        $author = $this->createTestUser('author', 'author@example.com', 'Author');
        $post = $this->createTestPost($author->pk_user, 'Original Title', 'Original Content');
        $postId = $post->id;
        $newTitle = 'Updated Title';
        $newContent = 'Updated Content';

        // Act
        $post->update([
            'title' => $newTitle,
            'content' => $newContent,
        ]);

        // Verify
        $response = $this->getJson("/api/posts/{$postId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson([
                'title' => $newTitle,
                'content' => $newContent,
            ]);
    }

    // ========================================================================
    // State Change Verification Tests
    // ========================================================================

    public function test_sequential_updates_accumulate(): void
    {
        // Arrange
        $user = $this->createTestUser('alice', 'alice@example.com', 'Alice');
        $userId = $user->id;

        // Act - first update
        $user->update(['bio' => 'Bio v1']);

        // Verify first change
        $response1 = $this->getJson("/api/users/{$userId}");
        $data1 = $response1->json();
        $this->assertEquals('Bio v1', $data1['bio']);

        // Second update
        $user->update(['full_name' => 'Alice Updated']);

        // Verify both changes accumulated
        $response2 = $this->getJson("/api/users/{$userId}");

        // Assert
        $response2->assertStatus(200)
            ->assertJson([
                'bio' => 'Bio v1',
                'fullName' => 'Alice Updated',
            ]);
    }

    public function test_update_one_user_does_not_affect_others(): void
    {
        // Arrange
        $user1 = $this->createTestUser('alice', 'alice@example.com', 'Alice');
        $user2 = $this->createTestUser('bob', 'bob@example.com', 'Bob');
        $newBio = "Alice's new bio";

        // Act
        $user1->update(['bio' => $newBio]);

        // Verify
        $response1 = $this->getJson("/api/users/{$user1->id}");
        $response2 = $this->getJson("/api/users/{$user2->id}");

        // Assert
        $response1->assertJson(['bio' => $newBio]);
        $data2 = $response2->json();
        $this->assertNull($data2['bio']);
    }

    public function test_update_one_post_does_not_affect_others(): void
    {
        // Arrange
        $author = $this->createTestUser('author', 'author@example.com', 'Author');
        $post1 = $this->createTestPost($author->pk_user, 'Post 1', 'Content 1');
        $post2 = $this->createTestPost($author->pk_user, 'Post 2', 'Content 2');
        $newTitle = 'Updated Title';

        // Act
        $post1->update(['title' => $newTitle]);

        // Verify
        $response1 = $this->getJson("/api/posts/{$post1->id}");
        $response2 = $this->getJson("/api/posts/{$post2->id}");

        // Assert
        $response1->assertJson(['title' => $newTitle]);
        $response2->assertJson(['title' => 'Post 2']);
    }

    // ========================================================================
    // Immutable Field Protection Tests
    // ========================================================================

    public function test_update_cannot_change_username(): void
    {
        // Arrange
        $user = $this->createTestUser('alice', 'alice@example.com', 'Alice');
        $userId = $user->id;
        $originalUsername = 'alice';

        // Act - Attempt to change username
        $user->update(['bio' => 'New bio']);  // Only update bio

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert - username should remain unchanged
        $response->assertJson(['username' => $originalUsername]);
    }

    // ========================================================================
    // Input Validation in Updates
    // ========================================================================

    public function test_update_user_with_special_characters(): void
    {
        // Arrange
        $user = $this->createTestUser('alice', 'alice@example.com', 'Alice');
        $userId = $user->id;
        $specialBio = "Bio with 'quotes', \"double quotes\", <html>, & ampersand";

        // Act
        $user->update(['bio' => $specialBio]);

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson(['bio' => $specialBio]);
    }

    public function test_update_user_with_unicode_characters(): void
    {
        // Arrange
        $user = $this->createTestUser('alice', 'alice@example.com', 'Alice');
        $userId = $user->id;
        $unicodeBio = "Bio with émojis 🎉 and spëcial chàrs";

        // Act
        $user->update(['bio' => $unicodeBio]);

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson(['bio' => $unicodeBio]);
    }

    public function test_update_post_content_with_long_text(): void
    {
        // Arrange
        $author = $this->createTestUser('author', 'author@example.com', 'Author');
        $post = $this->createTestPost($author->pk_user, 'Title', 'Short content');
        $postId = $post->id;
        $longContent = str_repeat('x', 5000);

        // Act
        $post->update(['content' => $longContent]);

        // Verify
        $response = $this->getJson("/api/posts/{$postId}");

        // Assert
        $response->assertStatus(200);
        $data = $response->json();
        $this->assertEquals(5000, strlen($data['content']));
    }

    public function test_update_user_bio_to_null(): void
    {
        // Arrange
        $user = $this->createTestUser('alice', 'alice@example.com', 'Alice', 'Old bio');
        $userId = $user->id;

        // Act
        $user->update(['bio' => null]);

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert
        $response->assertStatus(200);
        $data = $response->json();
        $this->assertNull($data['bio']);
    }

    public function test_update_user_bio_to_empty_string(): void
    {
        // Arrange
        $user = $this->createTestUser('alice', 'alice@example.com', 'Alice', 'Old bio');
        $userId = $user->id;

        // Act
        $user->update(['bio' => '']);

        // Verify
        $response = $this->getJson("/api/users/{$userId}");

        // Assert
        $response->assertStatus(200)
            ->assertJson(['bio' => '']);
    }
}

File: frameworks/ruby-rails/test/controllers/mutation_tests.rb
require "test_helper"

class MutationTests < ActionDispatch::IntegrationTest
  # ========================================================================
  # Single Field Update Tests
  # ========================================================================

  test "update user bio single field" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice", "Old bio")
    user_id = user.id
    new_bio = "Updated bio"

    # Act
    user.update(bio: new_bio)

    # Verify
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal new_bio, data["bio"]
    # Verify other fields unchanged
    assert_equal "Alice", data["fullName"]
  end

  test "update user full name single field" do
    # Arrange
    user = create_test_user("bob", "bob@example.com", "Bob")
    user_id = user.id
    new_name = "Bob Smith Updated"

    # Act
    user.update(full_name: new_name)

    # Verify
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal new_name, data["fullName"]
  end

  test "update post title" do
    # Arrange
    author = create_test_user("author", "author@example.com", "Author")
    post = create_test_post(author.pk_user, "Original Title", "Original Content")
    post_id = post.id
    new_title = "Updated Title"

    # Act
    post.update(title: new_title)

    # Verify
    get "/api/posts/#{post_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal new_title, data["title"]
    assert_equal "Original Content", data["content"]
  end

  test "update post content" do
    # Arrange
    author = create_test_user("author", "author@example.com", "Author")
    post = create_test_post(author.pk_user, "Title", "Original Content")
    post_id = post.id
    new_content = "Updated Content"

    # Act
    post.update(content: new_content)

    # Verify
    get "/api/posts/#{post_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal new_content, data["content"]
  end

  # ========================================================================
  # Multi-Field Update Tests
  # ========================================================================

  test "update user multiple fields" do
    # Arrange
    user = create_test_user("charlie", "charlie@example.com", "Charlie")
    user_id = user.id
    new_bio = "New bio"
    new_name = "Charlie Updated"

    # Act
    user.update(bio: new_bio, full_name: new_name)

    # Verify
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal new_bio, data["bio"]
    assert_equal new_name, data["fullName"]
  end

  test "update post title and content" do
    # Arrange
    author = create_test_user("author", "author@example.com", "Author")
    post = create_test_post(author.pk_user, "Original Title", "Original Content")
    post_id = post.id
    new_title = "Updated Title"
    new_content = "Updated Content"

    # Act
    post.update(title: new_title, content: new_content)

    # Verify
    get "/api/posts/#{post_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal new_title, data["title"]
    assert_equal new_content, data["content"]
  end

  # ========================================================================
  # State Change Verification Tests
  # ========================================================================

  test "sequential updates accumulate" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    user_id = user.id

    # Act - first update
    user.update(bio: "Bio v1")

    # Verify first change
    get "/api/users/#{user_id}"
    data1 = JSON.parse(@response.body)
    assert_equal "Bio v1", data1["bio"]

    # Second update
    user.update(full_name: "Alice Updated")

    # Verify both changes accumulated
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data2 = JSON.parse(@response.body)
    assert_equal "Bio v1", data2["bio"]
    assert_equal "Alice Updated", data2["fullName"]
  end

  test "update one user does not affect others" do
    # Arrange
    user1 = create_test_user("alice", "alice@example.com", "Alice")
    user2 = create_test_user("bob", "bob@example.com", "Bob")
    new_bio = "Alice's new bio"

    # Act
    user1.update(bio: new_bio)

    # Verify
    get "/api/users/#{user1.id}"
    response1 = @response.body
    get "/api/users/#{user2.id}"
    response2 = @response.body

    # Assert
    data1 = JSON.parse(response1)
    data2 = JSON.parse(response2)
    assert_equal new_bio, data1["bio"]
    assert_nil data2["bio"]
  end

  test "update one post does not affect others" do
    # Arrange
    author = create_test_user("author", "author@example.com", "Author")
    post1 = create_test_post(author.pk_user, "Post 1", "Content 1")
    post2 = create_test_post(author.pk_user, "Post 2", "Content 2")
    new_title = "Updated Title"

    # Act
    post1.update(title: new_title)

    # Verify
    get "/api/posts/#{post1.id}"
    response1 = @response.body
    get "/api/posts/#{post2.id}"
    response2 = @response.body

    # Assert
    data1 = JSON.parse(response1)
    data2 = JSON.parse(response2)
    assert_equal new_title, data1["title"]
    assert_equal "Post 2", data2["title"]
  end

  # ========================================================================
  # Input Validation in Updates
  # ========================================================================

  test "update user with special characters" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    user_id = user.id
    special_bio = "Bio with 'quotes', \"double quotes\", <html>, & ampersand"

    # Act
    user.update(bio: special_bio)

    # Verify
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal special_bio, data["bio"]
  end

  test "update user with unicode characters" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    user_id = user.id
    unicode_bio = "Bio with émojis 🎉 and spëcial chàrs"

    # Act
    user.update(bio: unicode_bio)

    # Verify
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal unicode_bio, data["bio"]
  end

  test "update post content with long text" do
    # Arrange
    author = create_test_user("author", "author@example.com", "Author")
    post = create_test_post(author.pk_user, "Title", "Short content")
    post_id = post.id
    long_content = "x" * 5000

    # Act
    post.update(content: long_content)

    # Verify
    get "/api/posts/#{post_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal 5000, data["content"].length
  end

  test "update user bio to null" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice", "Old bio")
    user_id = user.id

    # Act
    user.update(bio: nil)

    # Verify
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_nil data["bio"]
  end

  test "update user bio to empty string" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice", "Old bio")
    user_id = user.id

    # Act
    user.update(bio: "")

    # Verify
    get "/api/users/#{user_id}"

    # Assert
    assert_response :success
    data = JSON.parse(@response.body)
    assert_equal "", data["bio"]
  end
end
